File: frontend/src/lib/odMatrix.ts
// The OD spread of DEMAND 0.1.0 (docs/2026-09-18_manual_demand_guide.md
// D25–D30), ported one-to-one from the sketch's odShares(), pinPair() and
// presetWeights() and mirrored by the backend's models/demand/od_matrix.py.
// Shares are fractions summing to 1 over the SELLABLE pairs — a stop the
// trip boards at before a stop it alights at; night stops sell nothing —
// keyed by stop id, never by name. Pins are posted and shown in percent.

export interface SellablePair {
  originStopId: string
  destinationStopId: string
  distanceKm: number
}

export type PinMap = Record<string, Record<string, number>>

export const OD_PRESETS = ['even', 'long', 'mid', 'short'] as const
export type OdPreset = (typeof OD_PRESETS)[number]

/** A preset maps a stop's position factor f (0 first … 1 last of its list)
 *  to OD_WEIGHT_MIN + OD_WEIGHT_SPAN × f, one decimal; even is 1. */
export const OD_WEIGHT_MIN = 0.3
export const OD_WEIGHT_SPAN = 2.7
export const DEFAULT_OD_WEIGHT = 1

const PRESET_FACTORS: Record<OdPreset, [(x: number) => number, (x: number) => number]> = {
  even: [() => 1, () => 1],
  long: [(x) => 1 - x, (x) => x],
  mid: [(x) => 1 - Math.abs(2 * x - 1), (x) => 1 - Math.abs(2 * x - 1)],
  short: [(x) => x, (x) => 1 - x],
}

/** The sellable pairs of a compact trip in route order: origins outer,
 *  destinations inner. stops carry their stop_type as the backend classifies
 *  them; km are cumulative from the trip's segments. */
export function sellablePairs(
  stops: { stop_id: string; stop_type: string }[],
  segmentKm: number[],
): SellablePair[] {
  const km = [0]
  for (const d of segmentKm) km.push(km[km.length - 1] + d)
  const boards = (t: string) => t === 'BOARDING' || t === 'BOTH'
  const alights = (t: string) => t === 'ALIGHTING' || t === 'BOTH'
  const out: SellablePair[] = []
  for (let i = 0; i < stops.length; i++) {
    if (!boards(stops[i].stop_type)) continue
    for (let j = i + 1; j < stops.length; j++) {
      if (!alights(stops[j].stop_type)) continue
      out.push({
        originStopId: stops[i].stop_id,
        destinationStopId: stops[j].stop_id,
        distanceKm: km[j] - km[i],
      })
    }
  }
  return out
}

export function boardingStopIds(stops: { stop_id: string; stop_type: string }[]): string[] {
  return stops
    .filter((s) => s.stop_type === 'BOARDING' || s.stop_type === 'BOTH')
    .map((s) => s.stop_id)
}

export function alightingStopIds(stops: { stop_id: string; stop_type: string }[]): string[] {
  return stops
    .filter((s) => s.stop_type === 'ALIGHTING' || s.stop_type === 'BOTH')
    .map((s) => s.stop_id)
}

/** The stop weights a preset writes. */
export function presetWeights(
  boardIds: string[],
  alightIds: string[],
  preset: OdPreset,
): { board: Record<string, number>; alight: Record<string, number> } {
  const [boardF, alightF] = PRESET_FACTORS[preset]
  const pos = (ids: string[], i: number) => (ids.length > 1 ? i / (ids.length - 1) : 0.5)
  // toFixed(1) as the sketch does; the backend rounds to one decimal too, and
  // both round 1.6500000000000001 up.
  const weight = (f: number) =>
    preset === 'even' ? 1 : +(OD_WEIGHT_MIN + OD_WEIGHT_SPAN * f).toFixed(1)
  const board: Record<string, number> = {}
  const alight: Record<string, number> = {}
  boardIds.forEach((s, i) => (board[s] = weight(boardF(pos(boardIds, i)))))
  alightIds.forEach((s, i) => (alight[s] = weight(alightF(pos(alightIds, i)))))
  return { board, alight }
}

/** One weight per stop of the list: the given one where it names the stop,
 *  DEFAULT_OD_WEIGHT elsewhere. */
export function resolveWeights(
  ids: string[],
  weights: Record<string, number> | undefined,
): Record<string, number> {
  const out: Record<string, number> = {}
  for (const s of ids) out[s] = weights?.[s] ?? DEFAULT_OD_WEIGHT
  return out
}

export const pinKey = (o: string, d: string) => `${o}\u0000${d}`

function pinOf(pins: PinMap, p: SellablePair): number | undefined {
  return pins[p.originStopId]?.[p.destinationStopId]
}

/** Share per sellable pair, fractions summing to 1: board × alight weight,
 *  normalised over the unpinned pairs onto whatever the pins leave; a pinned
 *  pair keeps its value. Returned in the pairs' order. */
export function odShares(
  pairs: SellablePair[],
  board: Record<string, number>,
  alight: Record<string, number>,
  pins: PinMap = {},
): number[] {
  const base = pairs.map((p) => (board[p.originStopId] ?? 0) * (alight[p.destinationStopId] ?? 0))
  const pinned = pairs.reduce((t, p) => t + (pinOf(pins, p) ?? 0), 0)
  const free = pairs.reduce((t, p, i) => t + (pinOf(pins, p) === undefined ? base[i] : 0), 0)
  const rest = Math.max(0, 100 - pinned)
  return pairs.map((p, i) => {
    const pin = pinOf(pins, p)
    return (pin !== undefined ? pin : free ? (base[i] / free) * rest : 0) / 100
  })
}

/** Pin one pair at valuePct; the other pins rescale by (100 − v) / (100 − old)
 *  and the unpinned pairs through the remainder, so the total stays 100. */
export function pinPair(
  pairs: SellablePair[],
  board: Record<string, number>,
  alight: Record<string, number>,
  pins: PinMap,
  origin: string,
  destination: string,
  valuePct: number,
): PinMap {
  const current = odShares(pairs, board, alight, pins)
  const index = pairs.findIndex(
    (p) => p.originStopId === origin && p.destinationStopId === destination,
  )
  const old = index >= 0 ? current[index] * 100 : 0
  const value = Math.min(100, Math.max(0, valuePct))
  const k = old < 100 ? (100 - value) / (100 - old) : 0
  const next: PinMap = {}
  for (const [o, row] of Object.entries(pins)) {
    for (const [d, v] of Object.entries(row)) {
      if (o === origin && d === destination) continue
      ;(next[o] ??= {})[d] = +(v * k).toFixed(2)
    }
  }
  ;(next[origin] ??= {})[destination] = +value.toFixed(2)
  return next
}

/** Drop a pin — the pair rejoins the unpinned remainder. */
export function unpinPair(pins: PinMap, origin: string, destination: string): PinMap {
  const next: PinMap = {}
  for (const [o, row] of Object.entries(pins)) {
    const kept = Object.fromEntries(
      Object.entries(row).filter(([d]) => !(o === origin && d === destination)),
    )
    if (Object.keys(kept).length) next[o] = kept
  }
  return next
}

/** The share-weighted journey length. */
export function averageDistanceKm(pairs: SellablePair[], shares: number[]): number {
  return pairs.reduce((s, p, i) => s + p.distanceKm * (shares[i] ?? 0), 0)
}

File: backend/models/demand/od_matrix.py
"""
od_matrix.py
============
The spread of one train's sales over its OD pairs (docs/2026-09-18_manual_
demand_guide.md D25–D30): an origin × destination matrix of shares over
the SELLABLE pairs of a trip, filled from one weight per boarding and
alighting stop, with pinned cells that hold their share while the rest
rescales.

Sellable pair: a stop the trip boards at (BOARDING or BOTH) before a
stop it alights at (ALIGHTING or BOTH). NIGHT stops are on neither side
— demand-quiet by definition; they still dwell and cost, they just sell
no places. This is the rule the stopgap carried since ROUTE_BUILDER
0.9.13, now in one place.

Shares are keyed by (origin_stop_id, destination_stop_id) — stop ids,
never names — and expressed as fractions summing to 1 over the sellable
pairs (pins are posted and echoed in percent). One-to-one ports of the
sketch's odShares(), pinPair() and presetWeights(); the frontend's
lib/odMatrix.ts is the same arithmetic and tests/test_83_demand_units.py
pins both.

Public interface:
  SellablePair(origin_stop_id, destination_stop_id, distance_km)
  sellable_pairs(trip) -> list[SellablePair]           # route order
  boarding_stop_ids(trip), alighting_stop_ids(trip)    # route order
  preset_weights(board_ids, alight_ids, preset) -> (board, alight)
  resolve_weights(stop_ids, weights) -> dict           # defaults filled
  od_shares(pairs, board, alight, pins_pct) -> dict[(o, d), fraction]
  pin_pair(pairs, board, alight, pins_pct, key, value_pct) -> pins_pct
  drop_stale_pins(pins_pct, pairs) -> (kept, dropped)
  average_distance_km(pairs, shares) -> float
"""

from __future__ import annotations

from dataclasses import dataclass

from models.demand.model import (
    DEFAULT_OD_WEIGHT,
    OD_PRESETS,
    OD_WEIGHT_MIN,
    OD_WEIGHT_SPAN,
)
from models.route.trip import StopType, Trip

PairKey = tuple[str, str]


@dataclass(frozen=True)
class SellablePair:
    origin_stop_id: str
    destination_stop_id: str
    distance_km: float

    @property
    def key(self) -> PairKey:
        return (self.origin_stop_id, self.destination_stop_id)


def _cumulative_km(trip: Trip) -> list[float]:
    out = [0.0]
    for segment in trip.segments:
        out.append(out[-1] + segment.distance_m / 1000.0)
    return out


def boarding_stop_ids(trip: Trip) -> list[str]:
    return [
        s.stop_id
        for s in trip.stops
        if s.stop_type in (StopType.BOARDING, StopType.BOTH)
    ]


def alighting_stop_ids(trip: Trip) -> list[str]:
    return [
        s.stop_id
        for s in trip.stops
        if s.stop_type in (StopType.ALIGHTING, StopType.BOTH)
    ]


def sellable_pairs(trip: Trip) -> list[SellablePair]:
    """Every (boarding stop, later alighting stop) of the trip with its
    journey length, in route order — origins outer, destinations inner."""
    stops = trip.stops
    km = _cumulative_km(trip)
    return [
        SellablePair(stops[i].stop_id, stops[j].stop_id, km[j] - km[i])
        for i in range(len(stops))
        for j in range(i + 1, len(stops))
        if stops[i].stop_type in (StopType.BOARDING, StopType.BOTH)
        and stops[j].stop_type in (StopType.ALIGHTING, StopType.BOTH)
    ]


# The four fill presets as position factors (D27): x is the stop's position
# in its list, 0 = first, 1 = last. What the margins show is what fills the
# matrix — there is no separate distance curve.
_PRESET_FACTORS = {
    "even": (lambda x: 1.0, lambda x: 1.0),
    "long": (lambda x: 1.0 - x, lambda x: x),
    "mid": (lambda x: 1.0 - abs(2.0 * x - 1.0), lambda x: 1.0 - abs(2.0 * x - 1.0)),
    "short": (lambda x: x, lambda x: 1.0 - x),
}


def preset_weights(
    board_ids: list[str], alight_ids: list[str], preset: str
) -> tuple[dict[str, float], dict[str, float]]:
    """The stop weights a preset writes: OD_WEIGHT_MIN + OD_WEIGHT_SPAN x f,
    one decimal, over the position factor f of each stop — except `even`,
    which is 1 everywhere rather than the formula's 3.0 (finding 5)."""
    if preset not in OD_PRESETS:
        raise ValueError(
            f"Unknown OD preset '{preset}'. Supported: {list(OD_PRESETS)}."
        )
    board_f, alight_f = _PRESET_FACTORS[preset]

    def pos(ids: list[str], i: int) -> float:
        return i / (len(ids) - 1) if len(ids) > 1 else 0.5

    def weight(f: float) -> float:
        return 1.0 if preset == "even" else round(OD_WEIGHT_MIN + OD_WEIGHT_SPAN * f, 1)

    return (
        {s: weight(board_f(pos(board_ids, i))) for i, s in enumerate(board_ids)},
        {s: weight(alight_f(pos(alight_ids, i))) for i, s in enumerate(alight_ids)},
    )


def resolve_weights(
    stop_ids: list[str], weights: dict[str, float] | None
) -> dict[str, float]:
    """One weight per stop of the list, the request's where it names the
    stop and DEFAULT_OD_WEIGHT elsewhere; weights for stops not on the list
    are dropped (a stop that left the route keeps nothing here — the
    request echo is the place a client reads its weights back from)."""
    given = weights or {}
    return {s: float(given.get(s, DEFAULT_OD_WEIGHT)) for s in stop_ids}


def od_shares(
    pairs: list[SellablePair],
    board: dict[str, float],
    alight: dict[str, float],
    pins_pct: dict[PairKey, float] | None = None,
) -> dict[PairKey, float]:
    """Share of the demand per sellable pair, as fractions summing to 1:
    boarding weight x alighting weight, normalised over the unpinned pairs
    onto whatever the pins leave; a pinned pair keeps its value (D26,
    D28). All weights 0 and no pins → every pair 0."""
    pins = pins_pct or {}
    base = {
        p.key: board.get(p.origin_stop_id, 0.0) * alight.get(p.destination_stop_id, 0.0)
        for p in pairs
    }
    pinned = sum(pins.get(p.key, 0.0) for p in pairs)
    free = sum(base[p.key] for p in pairs if p.key not in pins)
    rest = max(0.0, 100.0 - pinned)
    return {
        p.key: (
            pins[p.key]
            if p.key in pins
            else (base[p.key] / free * rest if free else 0.0)
        )
        / 100.0
        for p in pairs
    }


def pin_pair(
    pairs: list[SellablePair],
    board: dict[str, float],
    alight: dict[str, float],
    pins_pct: dict[PairKey, float],
    key: PairKey,
    value_pct: float,
) -> dict[PairKey, float]:
    """Pin one pair at value_pct; every OTHER pair rescales so the total
    stays 100 — the other pins explicitly, by (100 − v) / (100 − old)
    (finding 6), the unpinned ones by themselves through the remainder.
    Returns the new pin map (percent, two decimals like the sketch)."""
    current = od_shares(pairs, board, alight, pins_pct)
    old = current.get(key, 0.0) * 100.0
    value = min(100.0, max(0.0, float(value_pct)))
    k = (100.0 - value) / (100.0 - old) if old < 100.0 else 0.0
    new_pins = {other: round(v * k, 2) for other, v in pins_pct.items() if other != key}
    new_pins[key] = round(value, 2)
    return new_pins


def drop_stale_pins(
    pins_pct: dict[PairKey, float], pairs: list[SellablePair]
) -> tuple[dict[PairKey, float], list[PairKey]]:
    """Keep the pins whose pair is still sellable on this trip; report the
    rest, so the response can say which pins were dropped (§4)."""
    sellable = {p.key for p in pairs}
    kept = {k: v for k, v in pins_pct.items() if k in sellable}
    dropped = [k for k in pins_pct if k not in sellable]
    return kept, dropped


def average_distance_km(
    pairs: list[SellablePair], shares: dict[PairKey, float]
) -> float:
    """The share-weighted journey length — what the sketch's What-follows
    panel prices a class's revenue on."""
    return sum(p.distance_km * shares.get(p.key, 0.0) for p in pairs)
